encoding/rs/encoder.go

```diff
@@ -44,21 +107,42 @@ type RsComputeDevice interface {
 // original data. When some systematic chunks are missing but identical parity chunk are
 // available, the receive can go through a Reed Solomon decoding to reconstruct the
 // original data.
-func NewEncoder(params encoding.EncodingParams, verbose bool) (*Encoder, error) {
-
+func (e *Encoder) newEncoder(params encoding.EncodingParams) (*ParametrizedEncoder, error) {
 	err := params.Validate()
 	if err != nil {
 		return nil, err
 	}
 
+	fs := e.CreateFFTSettings(params)
+
+	switch e.Config.BackendType {
+	case encoding.BackendDefault:
+		return e.createDefaultBackendEncoder(params, fs)
+	case encoding.BackendIcicle:
+		return e.createIcicleBackendEncoder(params, fs)
+	default:
+		return nil, fmt.Errorf("unsupported backend type: %v", e.Config.BackendType)
+	}
+}
+
+func (e *Encoder) CreateFFTSettings(params encoding.EncodingParams) *fft.FFTSettings {
 	n := uint8(math.Log2(float64(params.NumEvaluations())))
-	fs := fft.NewFFTSettings(n)
+	return fft.NewFFTSettings(n)
+}
+
+func (e *Encoder) createDefaultBackendEncoder(params encoding.EncodingParams, fs *fft.FFTSettings) (*ParametrizedEncoder, error) {
+	if e.Config.GPUEnable {
+		return nil, fmt.Errorf("GPU is not supported in default backend")
+	}
 
-	return &Encoder{
-		EncodingParams: params,
-		Fs:             fs,
-		verbose:        verbose,
-		NumRSWorker:    runtime.GOMAXPROCS(0),
+	return &ParametrizedEncoder{
```

Comment:
So when you say "backend", it looks a software/library that computes RS encoding? I.e. it decoupled device (CPU/GPU) and backend (the software that executes compute on device)

Reply:
Right now I consider backend as library that implements the cryptographic primitive like NTT, MSM. We have our own implementation of NTT borrowed from protolambda and MSM from gnark. 

It could be plausible that there exists libraries that implement the higher level primitives like multiproofs, commitment, RS encode and we can try to make our code more of a frontend that calls those. But for now we focus on the core primitives that give the speedup factor.